### DISCUSSION TITLE
Example modules and scripts: Which is which? How to run with docker?

### DISCUSSION THREAD

**Discussion by happycherrycake** [lfe/lfe#456] [Q&A] [answered]
I'm building lfe the first time. to reproduce:
```
docker pull lfex/lfe@sha256:16e31c5941e0b571de315d251f51e9ceb72005368a833c828c327b5c7b9ecb2f
docker run lfex/lfe lfe /opt/erlang/lfe/examples/simple-erl-exercises.lfe
```

In another non-docker build:
````
git clone https://github.com/rvirding/lfe
cd lfe
git checkout 0775432025f43cafbb7063b5923286bbd700cdf0
make
./bin/lfe ./examples/simple-erl-exercises.lfe
````
It prints "exception error: undefined function define-module/3", which is similar but slightly different.

The build was mostly successful besides these warning messages:
````
src/lfe_io_pretty.erl:23: Warning: export_all flag enabled - all functions will be exported
src/lfe_lint.erl:1501: Warning: function add_errors/3 is unused
src/lfe_codegen.erl:33: Warning: export_all flag enabled - all functions will be exported
src/lfe_macro_include.erl:306: Warning: variable 'Line' is unused
src/lfe_types.erl:37: Warning: export_all flag enabled - all functions will be exported
````

In the lfe REPL, some basic functionality is fine, like '1', '(cons 1 2)'. I haven't tried extensively though.

**Comment by oubiwann** [comment]
It's been waaaaay too long since I've worked with LFE :-( Feeling so rusty. But here are some thoughts that may be out of date:
* is `simple-erl-exercises.lfe` meant to be run from the command line?
* take a look at another example which offers usage ... I think I wrote the Church numerals one, way back ... it the top-level comments describe usage there with `slurp`
* or the top-level comments in the `ring.lfe` example: that one shows you the command line options to use to execute a specific function (`main`) in a specific module (`ring`)

The reason I bring these up is that, by itself, the `simple-erl-exercises.lfe` file is just a module definition, not an executable nor an LFE script. Oh, for an LFE script, check out `examples/sample-lfescript` ...

**Comment by oubiwann** [comment]
Hrm, looks like the examples doen't execute like they used to. I just tried the following:

```
$ docker run -it lfex/lfe bash
$ cd /opt/erlang/lfe/examples
$ ../bin/lfe -pa ../ebin -smp disable -noshell -run ring main 503 50000000
```
And got this:
```
lfe> {"init terminating in do_boot",{undef,[{ring,main,[["503","50000000"]],[]},{init,start_em,1,[]},{init,do_boot,3,[]}]}}
init terminating in do_boot ()

Crash dump is being written to: erl_crash.dump...done
```

However (still in the Docker container), I did get this to work:
```
$ mkdir examples/ebin
$ lfec -I include -o examples/ebin examples/ring.lfe
$ lfe -pa examples/ebin -smp disable -noshell -run ring main 503 50000000
```
Note that it does print out the banner, and after a few seconds, it cranks out the expected result: `292`

**Comment by oubiwann** [comment]
Maybe the difference is simply that the old default `make` target compiled the examples, and the latest doesn't?

**Comment by oubiwann** [comment, marked as answer]
Okay, I've put some stuff together in order to give you a better answer here, that doesn't require building LFE locally, etc. Checkout the `README` at this project:
 * https://github.com/lfex/dockerfiles

There are a whole new set of LFE docker images, based on the recent Erlang docker images, and you can treat these images like the LFE binary essentially (since they use `ENTRYPOINT`).

This should provide you with the steps necessary to confirm that you can run a module with a `main` function from the CLI as well as run the example scripts.

Some general guidelines on what is runnable via the LFE CLI (standalone binary or Docker image):
* If it contains a module (look for the `defmodule` form in the file), you'll either need to compile it or slurp it
* If you compile it, you'll be able to run it from the CLI only if 1) it doesn't take any args, or 2) there's a function that handles passed args, converts them appropriately, etc.
* If you slurp it, you'll need to be in the REPL (and you won't need to reference the module name)
* If the code doesn't contain a `defmodule` form, it's either a script or just some code you can slurp
* You can tell it's a script by looking at the shebang line (e.g., `#! /usr/bin/env lfe` or `#! /usr/bin/env lfescript`)

I'm going to mark this as answered, and I'll change the title of the ticket so that it's more useful/descriptive for future queries ...

**Comment by oubiwann** [comment]
We've just done an update of all the examples in the `develop` branch, and things are working nicely per the instructions at the top of each example.
